governator-legacy/src/main/java/com/netflix/governator/lifecycle/ClasspathScanner.java

```diff
@@ -114,38 +113,66 @@ public ClasspathScanner(Collection<String> basePackages, Collection<Class<? exte
 
     protected void doScanning(Collection<String> basePackages, Collection<Class<? extends Annotation>> annotations, Set<Class<?>> localClasses, Set<Constructor> localConstructors, Set<Method> localMethods, Set<Field> localFields)
     {
-        if ( basePackages.size() == 0 )
+        if ( basePackages.isEmpty() )
         {
             log.warn("No base packages specified - no classpath scanning will be done");
             return;
         }
         try
         {
-            List<Archive> archives = Lists.newArrayList();
             for ( String basePackage : basePackages )
             {
                 Enumeration<URL> resources = classLoader.getResources(basePackage.replace(".", "/"));
                 while ( resources.hasMoreElements() )
                 {
-                    URL thisUrl = resources.nextElement();
-                    if ( isJarURL(thisUrl))
+                    URL url = resources.nextElement();
+                    if ( isJarURL(url))
                     {
-                        archives.add(new JarArchive(classLoader, thisUrl));
+                        try
+                        {
+                            String jarPath = url.getFile();
+                            if ( jarPath.contains("!") )
+                            {
+                                jarPath = jarPath.substring(0, jarPath.indexOf("!"));
+                                url = new URL(jarPath);
+                            }
+
+                            JarFile jar = new JarFile(ClasspathUrlDecoder.toFile(url));
+                            for ( Enumeration<JarEntry> list = jar.entries(); list.hasMoreElements(); )
+                            {
+                                JarEntry entry = list.nextElement();
+                                if ( entry.getName().endsWith(".class") )
+                                {
+                                    AnnotationFinder finder = new AnnotationFinder(classLoader, annotations.toArray(new Class[annotations.size()]));
+                                    new ClassReader(jar.getInputStream(entry)).accept(finder, SKIP_CODE);
+
+                                    localClasses.addAll(finder.getAnnotatedClasses());
+                                    localMethods.addAll(finder.getAnnotatedMethods());
+                                    localConstructors.addAll(finder.getAnnotatedConstructors());
+                                    localFields.addAll(finder.getAnnotatedFields());
+                                }
+                            }
+                        }
+                        catch( IOException e )
+                        {
+                            throw new IllegalStateException(e);
```

Comment:
Thanks.  This also brought to my attention that I wasn't closing the JarFile resource.  Oops!